mailpoet/package.json

```diff
@@ -180,24 +182,38 @@
     "js-yaml": "^4.1.0",
     "jsdom": "^19.0.0",
     "json-loader": "^0.5.7",
+    "meow": "^9.0.0",
     "mocha": "^9.2.2",
+    "path-browserify": "^1.0.1",
     "phplint": "^2.0.5",
     "postcss-cli": "^9.1.0",
     "postcss-scss": "^4.0.3",
+    "prismjs": "^1.28.0",
     "sass": "^1.49.9",
     "sinon": "^13.0.1",
     "sinon-chai": "^3.7.0",
     "storybook-addon-performance": "^0.16.1",
     "string-replace-loader": "^3.1.0",
-    "stylelint": "^14.6.1",
+    "stylelint": "^14.9.1",
     "stylelint-order": "^5.0.0",
     "stylelint-scss": "^4.2.0",
     "terser-webpack-plugin": "^5.3.1",
+    "trim": "^1.0.1",
     "ts-loader": "^9.2.8",
     "typescript": "^4.6.3",
+    "url": "^0.11.0",
     "webpack": "^5.70.0",
     "webpack-cli": "^4.9.2",
     "webpack-manifest-plugin": "^5.0.0",
     "wp-types": "^3.59.1"
+  },
+  "overrides": {
+    "prismjs": "$prismjs",
+    "webpack": "$webpack",
+    "react": "$react",
+    "underscore": "$underscore",
+    "trim": "$trim",
+    "glob-parent": "$glob-parent",
+    "meow": "$meow"
```

Comment:
@samnajian I haven't used `overrides` like this before. If I understand it correctly this will force any package that's referencing one of these to use the versions we're specifying in `dependencies` or `devDependencies`, is that right?

Would this explain why so many lines have been removed in `package-lock.json` compared to the number added? Perhaps there were lots of packages that were depending on different versions of these packages, and now that they're all referencing the same exact versions there are much fewer dependencies in general?

<img width="483" alt="image" src="https://user-images.githubusercontent.com/8656640/175618866-6fbe70a9-4e69-47ce-98db-2298daf423cb.png">

I'm just wondering if it's really necessary to include all of these in `overrides` and if it might cause problems down the line. Although this seems to be working fine now, what happens if we add a different package in the future that depends on `webpack` or `react` and doesn't work with the version we're using?

Please correct me if I'm not understanding this correctly.

It seems like we would want to keep this list as small as possible to reduce unexpected side effects in the future. Maybe we could document somewhere in this file, or the commit message(s), the specific packages that require us to include these overrides? 

Maybe we'll be able to remove some overrides in the future but it would be somewhat difficult to know which are okay to remove without more documentation (unless I'm missing something, in which case my apologies!).

Reply:
Thanks for taking into it @johnolek.
Your analogy for the big number of changes is the package-lock.json is correct.

> I'm just wondering if it's really necessary to include all of these in overrides and if it might cause problems down the line. Although this seems to be working fine now, what happens if we add a different package in the future that depends on Webpack or React and doesn't work with the version we're using?

On one hand, most of the overrides ( other than underscore ) are made to mitigate issues arising from packages used by Storybook which means it's pretty safe for the production build, and also the overrides are basically forcing Storybook to use the versions we use for the main app. 
On the other hand, these overrides are done via reference to the versions installed in dependencies/devDependencies sections ( using variables instead of hard-coding versions ), so updating them in the future, for the most part, wouldn't mean that we need to touch the overrides as well.

I hope this explains the PR a bit better, let me know if you still think we'd need more supporting documentation and then maybe we can work on a solution.

Reply:
@samnajian Thanks for the explanation!

If the vulnerable packages related to storybook aren't included/used in the production build, is it necessary to override them at all right now?

Based on the following discussions, it looks like storybook will have a vulnerability-free dependency tree with version 7.0:
https://github.com/storybookjs/storybook/issues/14603
https://github.com/storybookjs/storybook/issues/15174

Looking at the current [storybook releases](https://github.com/storybookjs/storybook/releases), it looks like they're currently have version 7 alpha 6, so it probably won't be too long until we could just update to version 7.

It makes sense to me to use overrides for dependencies that are no longer maintained, like `backbone.supermodel`. That package hasn't seen an update in 7 years, and because it's depending on `underscore` with a version of [~1.5.2](https://github.com/tanqhnguyen/backbone.supermodel/blob/9780ba0c855cf16ca3900ead2027ec3974494432/package.json#L33) it's preventing us from updating to a newer minor version of `underscore`. Therefore our only viable option in that case is to use an override. Am I understanding that correctly?

Since storybook version 7 should no longer have any vulnerable dependencies in its tree, and because the vulnerabilities that currently exist should only ever affect our dev environment, it seems like it would be a better option to wait for that version.

But even if these dependencies are only used by storybook right now, isn't it possible that they might get referenced by some of our production dependencies in the future?

For example, could something like this happen?

1. We lock some package P to version `^6.0.1` in our overrides. Right now only storybook depends on it and nothing breaks by forcing the override.
2. Some months or years from now we update one of our prod dependencies Q, and it now depends on package P version `^7.0`.
3. Because of our overrides, package Q will be forced to use version `^6.0.1` instead of `^7.0`. This could cause some unexpected and difficult-to-debug behavior.

Perhaps that would be very unlikely, but it still seems like it would be best to minimize the risk of it happening at all, unless `npm` would tell us that something like this had happened.

Long story short, my impression is that the more overrides we have the more risk there is that we would see subtle problems in the future, and it could be difficult to know when it's safe to update those overrides. 

Please let me know if I'm misunderstanding anything and/or if you think there's a compelling reason to use overrides now for the storybook-related vulnerabilities instead of waiting to update to storybook version 7 when it's released.

Reply:
@johnolek thanks alot for a deep look into things.

> Long story short, my impression is that the more overrides we have the more risk there is that we would see subtle problems in the future, and it could be difficult to know when it's safe to update those overrides.

The first part is correct, but for the second part: later on for instance as we update the storybook ( per our usual recurring npm package update tickets like [MAILPOET-4257] ) we'll simply remove the overrides and run `npm audit` again which would tell us if we are green or not, the statement you made about future updates is always true and we should keep running npm audit after each npm package update and I don't think this ticket is going to change that.

> Please let me know if I'm misunderstanding anything and/or if you think there's a compelling reason to use overrides now for the storybook-related vulnerabilities instead of waiting to update to storybook version 7 when it's released.

I'd like to get rid of the vulnerabilities as soon as possible for obvious reasons and then enforce running npm audit after each package update and update the overrides since just like what you mentioned in a perfect world we should have no overrides at all and we can always drive towards perfection!